integration/container/attach_test.go

```diff
@@ -0,0 +1,206 @@
+			Stderr: true,
+		},
+	)
+	assert.NilError(t, err)
+	defer attach.Close()
+
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		time.Sleep(300 * time.Millisecond)
+		err = client.ContainerStop(ctx, cID, nil)
+		assert.NilError(t, err)
+		_, err = io.Copy(&bytes.Buffer{}, attach.Reader)
+		errCh <- err
+	}()
+
+	poll.WaitOn(t, container.IsInState(ctx, client, cID, "exited"), poll.WithDelay(5*time.Second))
+
+	select {
+	case err = <-errCh:
+		assert.NilError(t, err)
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting for attach channel to be closed")
+	}
+}
+
+func TestAttachAfterDetach(t *testing.T) {
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true),
+		container.WithCmd("sh"),
+		func(c *container.TestContainerConfig) {
+			c.Config.OpenStdin = true
+		},
+	)
+
+	defer client.ContainerRemove(context.Background(), cID, types.ContainerRemoveOptions{Force: true})
+	poll.WaitOn(t, container.IsInState(ctx, client, cID, "running"), poll.WithDelay(5*time.Second))
+
+	attach, err := client.ContainerAttach(ctx, cID,
+		types.ContainerAttachOptions{
+			Stream: true,
+			Stdin:  true,
+			Stdout: true,
+			Stderr: true,
+		},
+	)
+	assert.NilError(t, err)
+
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		_, err := io.Copy(&bytes.Buffer{}, attach.Reader)
+		errCh <- err
+	}()
+
+	_, err = attach.Conn.Write([]byte{16})
+	assert.NilError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	_, err = attach.Conn.Write([]byte{17})
+	assert.NilError(t, err)
+
+	select {
+	case err := <-errCh:
+		assert.NilError(t, err)
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out while detaching")
+	}
+	attach.Close()
+
+	attach, err = client.ContainerAttach(ctx, cID,
+		types.ContainerAttachOptions{
+			Stream: true,
+			Stdin:  true,
+			Stdout: true,
+			Stderr: true,
+		},
+	)
+	assert.NilError(t, err)
+	defer attach.Close()
+
+	buf := make([]byte, 10)
+	var nBytes int
+	errCh = make(chan error)
+	go func() {
+		defer close(errCh)
+		_, err := attach.Conn.Write([]byte("\n"))
+		assert.NilError(t, err)
+		time.Sleep(500 * time.Millisecond)
+
+		nBytes, err = attach.Reader.Read(buf)
+		errCh <- err
+	}()
+
+	select {
+	case err := <-errCh:
+		assert.NilError(t, err)
+	case <-time.After(2 * time.Second):
+		t.Fatal("timed out waiting for attach read")
+	}
+
+	assert.Assert(t, is.Contains(string(buf[:nBytes]), "/ #"))
+}
+
+// TestAttachDetach checks that attach in tty mode can be detached using the long container ID
+func TestAttachDetach(t *testing.T) {
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true),
+		container.WithCmd("cat"),
+		func(c *container.TestContainerConfig) {
+			c.Config.OpenStdin = true
+		},
+	)
+	defer client.ContainerRemove(context.Background(), cID, types.ContainerRemoveOptions{Force: true})
```

Comment:
Same here as well I think the `defer setupTest(t)()` should already take care of this (if it isn't, then we need to fix that)